presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -381,12 +386,68 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
-            result.put(splitIndex % taskCount, splits.get(splitIndex));
+
+        long maxSplitsSizeInBytesPerPartition = getMaxSplitsDataSizePerSparkPartition(session).toBytes();
+        checkArgument(maxSplitsSizeInBytesPerPartition > 0,
+                "maxSplitsSizeInBytesPerPartition must be greater than zero: %s", maxSplitsSizeInBytesPerPartition);
+
+        boolean splitsDataSizeAvailable = splits.stream()
+                .allMatch(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().isPresent());
+        boolean autoTuneInitialPartitionCount = isSparkInitialPartitionCountAutoTuneEnabled(session) && splitsDataSizeAvailable;
+
+        int initialPartitionCount;
+        if (!autoTuneInitialPartitionCount) {
+            initialPartitionCount = getSparkInitialPartitionCount(session);
+            checkArgument(initialPartitionCount > 0, "initialPartitionCount must be greater then zero: %s", initialPartitionCount);
+        }
+        else {
+            long totalSizeInBytes = splits.stream()
+                    .mapToLong(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong())
+                    .sum();
+            initialPartitionCount = max(1, toIntExact(totalSizeInBytes + maxSplitsSizeInBytesPerPartition - 1 / maxSplitsSizeInBytesPerPartition));
+        }
+
+        if (!splitsDataSizeAvailable) {
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                result.put(splitIndex % initialPartitionCount, splits.get(splitIndex));
+            }
+        }
+        else {
+            splits.sort((ScheduledSplit o1, ScheduledSplit o2) ->
+                    o1.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() >
+                            o2.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() ? -1 : 1);
+
+            PriorityQueue<SparkPartition> queue = new PriorityQueue();
+            int partitionCount = 0;
+
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                int partitionId;
+                long splitSizeInBytes = splits.get(splitIndex).getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong();
+
+                if (partitionCount >= initialPartitionCount &&
+                        !queue.isEmpty() &&
+                        queue.peek().getAvailableCapacityInBytes() >= splitSizeInBytes) {
+                    SparkPartition partition = queue.poll();
+                    partitionId = partition.getPartitionId();
+                    partition.assignSplitWithSize(splitSizeInBytes);
+                    if (partition.getAvailableCapacityInBytes() > 0) {
+                        queue.add(partition);
+                    }
+                }
+                else {
+                    partitionCount++;
+                    partitionId = partitionCount - 1;
+                    if (splitSizeInBytes < maxSplitsSizeInBytesPerPartition) {
```

Comment:
What if split is larger? Does it mean that we are not going to schedule it? We are going to break correctness.

Reply:
This last line of code inside for loop guarantees each split will be assigned to a partition decided by variable `partitionId`.

Reply:
Oh, nevermind. I missread

Reply:
So basically if the split size is larger than the split size per partition theres no need to add the partition to the queue, as it will get full after adding this split, right?